crates/astria-composer/src/searcher/mod.rs

```diff
@@ -0,0 +1,117 @@
+use std::{
+    net::SocketAddr,
+    sync::Arc,
+};
+
+use thiserror::Error;
+use tokio::task::JoinSet;
+use tracing::{
+    error,
+    info,
+};
+
+use self::api::ApiError;
+use crate::config::searcher::{
+    Config,
+    ConfigError,
+};
+
+mod api;
+
+#[derive(Debug, Error)]
+pub enum SearcherError {
+    #[error("invalid config")]
+    InvalidConfig(#[from] ConfigError),
+    #[error("api error")]
+    ApiError(#[from] ApiError),
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct State(Arc<()>);
+
+pub struct Searcher {
+    state: State,
+    api_url: SocketAddr,
+}
+
+impl Searcher {
+    pub fn new(config: Config) -> Result<Self, SearcherError> {
+        // configure rollup tx collector
+        // configure rollup tx bundler
+        // configure rollup tx executor
+
+        // init searcher state
+        let state = State(Arc::new(()));
+
+        // parse api url from config and init api router
+        let api_url = config
```

Comment:
I know this is not how it's done in the other crates of this repo (although the plan is to slowly change that).

But if you do this whole thing from scratch, just create a `Config` that already contains the final types that you expect. So if you know that `api_url` must be a `SocketAddr`, then just set it to that type in `Config` and let figment/serde handle that.